kube-proxy panic because of service sessionAffinityConfig is nil

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug


**What happened**:

upgrading k8s to latest version, but keep the data in etcd, especially the services which session affinity type is ClientIP. Then kube-proxy panic because of service sessionAffinityConfig is nil.





**What you expected to happen**:

backward compatible with latency services.

**How to reproduce it (as minimally and precisely as possible)**:

For example,

```
╰─k8s.io/kubernetes> k get service kubernetes -o json
{
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
        "creationTimestamp": "2017-06-28T16:05:11Z",
        "labels": {
            "component": "apiserver",
            "provider": "kubernetes"
        },
        "name": "kubernetes",
        "namespace": "default",
        "resourceVersion": "6",
        "selfLink": "/api/v1/namespaces/default/services/kubernetes",
        "uid": "909a1c62-5c1b-11e7-8a60-4ccc6ab9973d"
    },
    "spec": {
        "clusterIP": "10.254.0.1",
        "ports": [
            {
                "name": "https",
                "port": 443,
                "protocol": "TCP",
                "targetPort": 6443
            }
        ],
        "sessionAffinity": "ClientIP",
        "type": "ClusterIP"
    },
    "status": {
        "loadBalancer": {}
    }
}
```

See error logs:

/data/go/src/runtime/asm_amd64.s:2197
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1616744]

goroutine 48 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
panic(0x18b5200, 0x2610290)
        /data/go/src/runtime/panic.go:489 +0x2cf
k8s.io/kubernetes/pkg/proxy/iptables.newServiceInfo(0xc4203bc7f0, 0x7, 0xc4203bc7e0, 0xa, 0xc4203bc82a, 0x5, 0xc42075c730, 0xc4203805a0, 0x9eb318c987fc4ed7)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:199 +0x9d4
k8s.io/kubernetes/pkg/proxy/iptables.serviceToServiceMap(0xc4203805a0, 0xc4206e2060)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:875 +0x1c8
k8s.io/kubernetes/pkg/proxy/iptables.(*serviceChangeMap).update(0xc420385620, 0xc420489df8, 0x0, 0xc4203805a0, 0x1ad7000)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:312 +0xb4
k8s.io/kubernetes/pkg/proxy/iptables.(*Proxier).OnServiceAdd(0xc420385600, 0xc4203805a0)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:645 +0x9a
k8s.io/kubernetes/pkg/proxy/config.(*ServiceConfig).handleAddService(0xc420295ad0, 0x1a807c0, 0xc4203805a0)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/config/config.go:225 +0xe8
k8s.io/kubernetes/pkg/proxy/config.(*ServiceConfig).(k8s.io/kubernetes/pkg/proxy/config.handleAddService)-fm(0x1a807c0, 0xc4203805a0)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/config/config.go:182 +0x3e
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(0xc42042ae90, 0xc42042aea0, 0xc42042aec0, 0x1a807c0, 0xc4203805a0)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:195 +0x49
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc4204e2420, 0x1a807c0, 0xc4203805a0)
        <autogenerated>:57 +0x73
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc4202fd130)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:545 +0x287
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).(k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.run)-fm()
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:381 +0x2a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc42000fa98, 0xc420202160)
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
        /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x62

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/cc @gnufied, thanks for reporting this.

--- a/pkg/api/fuzzer/fuzzer.go
+++ b/pkg/api/fuzzer/fuzzer.go
@@ -439,6 +439,19 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 					ss.Ports[i].TargetPort.StrVal = "x" + ss.Ports[i].TargetPort.StrVal // non-empty
 				}
 			}
+			types := []api.ServiceAffinity{api.ServiceAffinityNone, api.ServiceAffinityClientIP}
+			ss.SessionAffinity = types[c.Rand.Intn(len(types))]
+			switch ss.SessionAffinity {
+			case api.ServiceAffinityClientIP:
+				timeoutSeconds := int32(c.Rand.Intn(int(api.MaxClientIPServiceAffinitySeconds)))
+				ss.SessionAffinityConfig = &api.SessionAffinityConfig{
+					ClientIP: &api.ClientIPConfig{
+						TimeoutSeconds: &timeoutSeconds,
+					},
+				}
+			case api.ServiceAffinityNone:
+				ss.SessionAffinityConfig = nil
+			}
 		},
 		func(n *api.Node, c fuzz.Continue) {
 			c.FuzzNoCustom(n)

--- a/pkg/api/v1/defaults.go
+++ b/pkg/api/v1/defaults.go
@@ -101,6 +101,19 @@ func SetDefaults_Service(obj *v1.Service) {
 	if obj.Spec.SessionAffinity == "" {
 		obj.Spec.SessionAffinity = v1.ServiceAffinityNone
 	}
+	if obj.Spec.SessionAffinity == v1.ServiceAffinityNone {
+		obj.Spec.SessionAffinityConfig = nil
+	}
+	if obj.Spec.SessionAffinity == v1.ServiceAffinityClientIP {
+		if obj.Spec.SessionAffinityConfig == nil || obj.Spec.SessionAffinityConfig.ClientIP == nil || obj.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds == nil {
+			timeoutSeconds := v1.DefaultClientIPServiceAffinitySeconds
+			obj.Spec.SessionAffinityConfig = &v1.SessionAffinityConfig{
+				ClientIP: &v1.ClientIPConfig{
+					TimeoutSeconds: &timeoutSeconds,
+				},
+			}
+		}
+	}
 	if obj.Spec.Type == "" {
 		obj.Spec.Type = v1.ServiceTypeClusterIP
 	}

--- a/pkg/api/v1/defaults_test.go
+++ b/pkg/api/v1/defaults_test.go
@@ -667,8 +667,50 @@ func TestSetDefaultService(t *testing.T) {
 	if svc2.Spec.SessionAffinity != v1.ServiceAffinityNone {
 		t.Errorf("Expected default session affinity type:%s, got: %s", v1.ServiceAffinityNone, svc2.Spec.SessionAffinity)
 	}
+	if svc2.Spec.SessionAffinityConfig != nil {
+		t.Errorf("Expected empty session affinity config when session affinity type: %s, got: %v", v1.ServiceAffinityNone, svc2.Spec.SessionAffinityConfig)
+	}
 	if svc2.Spec.Type != v1.ServiceTypeClusterIP {
 		t.Errorf("Expected default type:%s, got: %s", v1.ServiceTypeClusterIP, svc2.Spec.Type)
+	}
+}
+
+func TestSetDefaultServiceSessionAffinityConfig(t *testing.T) {
+	testCases := map[string]v1.Service{
+		"SessionAffinityConfig is empty": {
+			Spec: v1.ServiceSpec{
+				SessionAffinity:       v1.ServiceAffinityClientIP,
+				SessionAffinityConfig: nil,
+			},
+		},
+		"ClientIP is empty": {
+			Spec: v1.ServiceSpec{
+				SessionAffinity: v1.ServiceAffinityClientIP,
+				SessionAffinityConfig: &v1.SessionAffinityConfig{
+					ClientIP: nil,
+				},
+			},
+		},
+		"TimeoutSeconds is empty": {
+			Spec: v1.ServiceSpec{
+				SessionAffinity: v1.ServiceAffinityClientIP,
+				SessionAffinityConfig: &v1.SessionAffinityConfig{
+					ClientIP: &v1.ClientIPConfig{
+						TimeoutSeconds: nil,
+					},
+				},
+			},
+		},
+	}
+	for name, test := range testCases {
+		obj2 := roundTrip(t, runtime.Object(&test))
+		svc2 := obj2.(*v1.Service)
+		if svc2.Spec.SessionAffinityConfig == nil || svc2.Spec.SessionAffinityConfig.ClientIP == nil || svc2.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds == nil {
+			t.Fatalf("Case: %s, unexpected empty SessionAffinityConfig/ClientIP/TimeoutSeconds when session affinity type: %s, got: %v", name, v1.ServiceAffinityClientIP, svc2.Spec.SessionAffinityConfig)
+		}
+		if *svc2.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds != v1.DefaultClientIPServiceAffinitySeconds {
+			t.Errorf("Case: %s, default TimeoutSeconds should be %d when session affinity type: %s, got: %d", name, v1.DefaultClientIPServiceAffinitySeconds, v1.ServiceAffinityClientIP, *svc2.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
+		}
 	}
 }
 

--- a/pkg/master/controller.go
+++ b/pkg/master/controller.go
@@ -250,7 +250,6 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
 		}
 		return nil
 	}
-	timeoutSeconds := api.DefaultClientIPServiceAffinitySeconds
 	svc := &api.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      serviceName,
@@ -264,11 +263,6 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
 			ClusterIP:       serviceIP.String(),
 			SessionAffinity: api.ServiceAffinityClientIP,
 			Type:            serviceType,
-			SessionAffinityConfig: &api.SessionAffinityConfig{
-				ClientIP: &api.ClientIPConfig{
-					TimeoutSeconds: &timeoutSeconds,
-				},
-			},
 		},
 	}
 

--- a/pkg/master/controller_test.go
+++ b/pkg/master/controller_test.go
@@ -546,7 +546,6 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 	om := func(name string) metav1.ObjectMeta {
 		return metav1.ObjectMeta{Namespace: ns, Name: name}
 	}
-	timeoutSeconds := api.DefaultClientIPServiceAffinitySeconds
 
 	create_tests := []struct {
 		testName     string
@@ -571,12 +570,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -631,12 +625,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -648,12 +637,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -674,12 +658,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -692,12 +671,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -717,12 +691,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -734,12 +703,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -759,12 +723,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -776,12 +735,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -801,12 +755,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -818,12 +767,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -843,12 +787,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -860,12 +799,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -885,12 +819,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeNodePort,
+					Type:            api.ServiceTypeNodePort,
 				},
 			},
 			expectUpdate: &api.Service{
@@ -902,12 +831,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 		},
@@ -927,12 +851,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: nil,
@@ -991,12 +910,7 @@ func TestCreateOrUpdateMasterService(t *testing.T) {
 					Selector:        nil,
 					ClusterIP:       "1.2.3.4",
 					SessionAffinity: api.ServiceAffinityClientIP,
-					SessionAffinityConfig: &api.SessionAffinityConfig{
-						ClientIP: &api.ClientIPConfig{
-							TimeoutSeconds: &timeoutSeconds,
-						},
-					},
-					Type: api.ServiceTypeClusterIP,
+					Type:            api.ServiceTypeClusterIP,
 				},
 			},
 			expectUpdate: nil,

--- a/pkg/proxy/iptables/proxier.go
+++ b/pkg/proxy/iptables/proxier.go
@@ -196,6 +196,7 @@ func newServiceInfo(svcPortName proxy.ServicePortName, port *api.ServicePort, se
 	}
 	var stickyMaxAgeSeconds int
 	if service.Spec.SessionAffinity == api.ServiceAffinityClientIP {
+		// Kube-apiserver side guarantees SessionAffinityConfig won't be nil when session affinity type is ClientIP
 		stickyMaxAgeSeconds = int(*service.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
 	}
 	info := &serviceInfo{

--- a/pkg/proxy/userspace/proxier.go
+++ b/pkg/proxy/userspace/proxier.go
@@ -448,7 +448,7 @@ func (proxier *Proxier) mergeService(service *api.Service) sets.String {
 		info.loadBalancerStatus = *helper.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
 		info.nodePort = int(servicePort.NodePort)
 		info.sessionAffinityType = service.Spec.SessionAffinity
-		// Set session affinity timeout value when sessionAffinity==ClientIP
+		// Kube-apiserver side guarantees SessionAffinityConfig won't be nil when session affinity type is ClientIP
 		if service.Spec.SessionAffinity == api.ServiceAffinityClientIP {
 			info.stickyMaxAgeSeconds = int(*service.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
 		}

--- a/staging/src/k8s.io/api/core/v1/types.go
+++ b/staging/src/k8s.io/api/core/v1/types.go
@@ -2998,6 +2998,8 @@ const (
 	ServiceAffinityNone ServiceAffinity = "None"
 )
 
+const DefaultClientIPServiceAffinitySeconds int32 = 10800
+
 // SessionAffinityConfig represents the configurations of session affinity.
 type SessionAffinityConfig struct {
 	// clientIP contains the configurations of Client IP based session affinity.

